### PR TITLE
#25 adaptive execution work without RepartitionByExpression action

### PR DESCRIPTION
**Pull request by shuiqing301**
## What changes were proposed in this pull request?

(Please fill in changes proposed in this fix)

## How was this patch tested?

(Please explain how this patch was tested. E.g. unit tests, integration tests, manual tests)
(If this patch involves UI changes, please attach a screenshot; otherwise, remove this)

Please review http://spark.apache.org/contributing.html before opening a pull request.

### CODE REVIEW

File: sql/core/src/main/scala/org/apache/spark/sql/execution/exchange/ShuffleExchangeExec.scala [hunk 1]
```diff
@@ -39,10 +39,10 @@ import org.apache.spark.util.collection.unsafe.sort.{PrefixComparators, RecordCo
...
  * Performs a shuffle that will result in the desired `newPartitioning`.
  */
-case class ShuffleExchangeExec(
-    var newPartitioning: Partitioning,
-    child: SparkPlan,
-    @transient coordinator: Option[ExchangeCoordinator]) extends Exchange {
+case class ShuffleExchangeExec(var newPartitioning: Partitioning,
+   child: SparkPlan,
+   isGenerated: Option[Boolean] = Some(true),
+   @transient coordinator: Option[ExchangeCoordinator]) extends Exchange {
```
Comment (hn5092) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/exchange/ShuffleExchangeExec.scala]: isGenerated这个把它放到 coordinator后面吧

File: core/src/test/scala/org/apache/spark/ui/UISeleniumSuite.scala [hunk 2]
```diff
@@ -265,6 +265,7 @@ class UISeleniumSuite extends SparkFunSuite with WebBrowser with Matchers with B
     }
   }
 
+  // scalastyle:off
```
Comment (hn5092) [inline core/src/test/scala/org/apache/spark/ui/UISeleniumSuite.scala]: 这里删了

File: sql/core/src/main/scala/org/apache/spark/sql/execution/exchange/EnsureRequirements.scala [hunk 3]
```diff
@@ -85,9 +86,9 @@ case class EnsureRequirements(conf: SQLConf) extends Rule[SparkPlan] {
             targetPostShuffleInputSize,
             minNumPostShufflePartitions)
         children.zip(requiredChildDistributions).map {
-          case (e: ShuffleExchangeExec, _) =>
+          case (e @ (ShuffleExchangeExec(_, _, _, Some(true))), _) =>
```
Comment (hn5092) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/exchange/EnsureRequirements.scala]: 括号删了

**Comment by shuiqing301** [comment]
retest this please

**Comment by shuiqing301** [comment]
retest this please

**Comment by shuiqing301** [comment]
retest this please

File: sql/core/src/main/scala/org/apache/spark/sql/execution/SparkStrategies.scala [hunk 4]
```diff
@@ -628,7 +628,7 @@ abstract class SparkStrategies extends QueryPlanner[SparkPlan] {
       case r: logical.Range =>
         execution.RangeExec(r) :: Nil
       case r: logical.RepartitionByExpression =>
-        exchange.ShuffleExchangeExec(r.partitioning, planLater(r.child)) :: Nil
+        exchange.ShuffleExchangeExec(r.partitioning, planLater(r.child), None, Some(false)) :: Nil
```
Comment (zheniantoushipashi) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/SparkStrategies.scala]: 感觉这个搞不太好，这个只要调用repartition 都丧失了  AE 的好处，我感觉这边应该加个配置

Comment (zheniantoushipashi) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/SparkStrategies.scala]: 几种调优方式，都是使用增多或者变少分区数来实现的, 在 repartition 的时候，跟这个实现有冲突, 这个 commit 忽略

**Comment by shuiqing301** [comment]
retest this please

**Comment by shuiqing301** [comment]
retest this please

**Comment by shuiqing301** [comment]
retest this please

**Pull request merged**